opm/parser/eclipse/EclipseState/Schedule/SegmentSet.cpp

```diff
@@ -0,0 +1,277 @@
+        } else {
+            return -1;
+        }
+    }
+
+    void SegmentSet::addSegment(SegmentPtr new_segment) {
+        // decide whether to push_back or insert
+        int segment_number = new_segment->segmentNumber();
+
+        int segment_location; // storage location
+        bool find = numberToLocation(segment_number, segment_location);
+
+        if (!find) { // it is a new segment
+            m_segments.push_back(new_segment);
+            m_number_to_location[segment_number] = m_number_segment;
+            ++ m_number_segment;
+        } else { // the segment already exists
+            m_segments[segment_location] = new_segment;
+        }
+    }
+
+    void SegmentSet::addSegmentINC(SegmentPtr /* new_segment */) {
+        throw std::logic_error("SegmentSet::addSegmentINC() not implemented yet");
+    }
+
+    void SegmentSet::addSegmentABS(SegmentPtr /* new_segment */) {
+        throw std::logic_error("SegmentSet::addSegmentABS() not implemented yet");
+    }
+
+    SegmentSet* SegmentSet::shallowCopy() const {
+        SegmentSet* copy = new SegmentSet();
+        copy->m_well_name = m_well_name;
+        copy->m_number_branch = m_number_branch;
+        copy->m_number_segment = m_number_segment;
+        copy->m_depth_top = m_depth_top;
+        copy->m_length_top = m_length_top;
+        copy->m_volume_top = m_volume_top;
+        copy->m_length_depth_type = m_length_depth_type;
+        copy->m_comp_pressure_drop = m_comp_pressure_drop;
+        copy->m_multiphase_model = m_multiphase_model;
+        copy->m_x_top = m_x_top;
+        copy->m_y_top = m_y_top;
+        copy->m_number_to_location = m_number_to_location;
+        copy->m_segments.resize(m_segments.size());
+        for (int i = 0; i < int(m_segments.size()); ++i) {
+            copy->m_segments[i] = m_segments[i];
+        }
+        return copy;
+    }
+
+    void SegmentSet::segmentsFromWELSEGSKeyword(DeckKeywordConstPtr welsegsKeyword, DeckKeywordConstPtr nsegmxKeyword) {
+        // get the required information for multi-segment wells
+        const int nsegmx = nsegmxKeyword->getRecord(0)->getItem("NSEGMX")->getInt(0);
+        const int nlbrmx = nsegmxKeyword->getRecord(0)->getItem("NLBRMX")->getInt(0);
+
+        // for the first record, which provides the information for the top segment
+        // and information for the whole segment set
+        DeckRecordConstPtr record1 = welsegsKeyword->getRecord(0);
+        m_well_name = record1->getItem("WELL")->getTrimmedString(0);
+
+        // create a temporary vector to store all the segments information in a sparse way
+        // then compress the vector in a orderly way to m_segments
+        std::vector<SegmentPtr> new_segments;
+        new_segments.resize(nsegmx);
+
+        for (int i = 0; i < int(new_segments.size()); ++i) {
+            new_segments[i] = NULL;
+        }
+
+        const double meaningless_value = -1.e100; // meaningless value to indicate unspecified values
```

Comment:
I *really* dislike such meaningless values - and I am willing to suggest quite significant rewrites to avoid it - although I do not see how just now.